libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -134,8 +134,31 @@ string IRGenerationContext::trySuccessConditionVariable(Expression const& _expre
 
 string IRGenerationContext::internalDispatch(Arity const& _arity)
 {
-	string funName = "dispatch_internal_in_" + to_string(_arity.in) + "_out_" + to_string(_arity.out);
-	return m_functions.createFunction(funName, [&]() {
+	// UNIMPLEMENTED: Internal library calls via pointers are not implemented yet.
+	// We're not generating code for internal library functions here even though it's possible
+	// to call them via pointers. Right now such calls end up triggering the `default` case in
+	// the switch above.
+	set<FunctionDefinition const*> functions;
+	for (auto const& contract: mostDerivedContract().annotation().linearizedBaseContracts)
+		for (FunctionDefinition const* function: contract->definedFunctions())
+			if (
+				FunctionType const* functionType = TypeProvider::function(*function)->asCallableFunction(false);
+				!function->isConstructor() &&
+				TupleType(functionType->parameterTypes()).sizeOnStack() == _arity.in &&
+				TupleType(functionType->returnParameterTypes()).sizeOnStack() == _arity.out
+			)
+			{
+				functions.insert(function);
+				enqueueFunctionForCodeGeneration(*function);
+			}
+
+	return internalDispatch(_arity, functions);
+}
+
+string IRGenerationContext::internalDispatch(Arity const& _arity, set<FunctionDefinition const*> const& _functions)
+{
+	string funName = IRNames::internalDispatch(_arity);
+	return m_functions.createFunction(funName, [&, functions(move(_functions))]() {
```

Comment:
Continuing discussion from https://github.com/ethereum/solidity/pull/8943#discussion_r425242170

> The lambda would iterate over `_functions` which is passed as a `const` reference from the outside. If `generateInternalDispatchFunction()` gets called by its other overload, the reference points at its local variable called `functions`. That variable no longer exists when the lambda eventually gets called.
>
> In #8797 this changes and the the reference will be pointing at `IRGenerationContext::m_internalDispatchMap` instead which we know will not go out of scope but it's just a happy coincidence - we don't do anything specific to ensure that.

Reply:
Actually `_arity` and `funName` also need to be passed by value. As far as I know, capturing a local variable by reference in a lambda that is not executed in the context of the current function is undefined behavior.

I just pushed a fix for this.